### ISSUE TITLE
Thoughts about long vectors

### ISSUE THREAD

**Issue by lemaitre** [WebAssembly/flexible-vectors#2]
Hello everyone,

Maybe it's a bit soon to start talking about long vectors where short vectors aren't specified yet, but I wanted to share my point of view on the subject.

Fixed width SIMD would be a nightmare to handle portably as not all the architectures have the same SIMD width.
For instance, AVX provides 256 bit registers, and AVX512 provides 512 bit registers.
Actually, ARM presented a while ago its new SIMD ISA: Scalable Vector Extension (SVE).

The idea behind SVE is that the size of the registers is not known during compilation and is hardware dependent: the same binary would be able to run efficiently on different hardware with different SIMD width.
I think that is exactly the abstraction we need for WASM to handle long vectors.
The SIMD width would be a constant that is not known at compile time (when the WASM code is generated), but known when the VM runs the code.

What do you think?

**Comment by lemaitre** [comment]
I don't get you, then...
I said the discussions were before SVE, and you said no... Maybe a quoting issue?
Could you rephrase to avoid any ambiguity?

**Comment by baryluk** [comment]
More semantic oriented vector types, that allow a loop or code to be specialized either to short or wide vectors (depending what hardware provides), or the ones that are flexible (i.e. depending on hardware), like ARM SVE, would really make it nice to use, and make the code even more portable and future proof I think. Even if the hardware doesn't support scalable vectors, compiler would know what it the wides available vector that can support all used operations and use it in generated code.

The big piece of it to abstract a width of a vector. I.e. provide reduce operations (`FloatWidest.sum`, `min`, etc) and allow looping to be width independent (i.e. `i += Float32Widest.size()`, and it will evaluate to constant determined at compile time).

I think design should really focus on AVX2/AVX512 and SVE in principle (and provide efficient fallbacks for "legacy" SIMD), even if these are not widely available yet. It will change in few years dramatically, and it is important for the Web. So for example, it should do use masking extensively, conflict detection, etc.

I understand wide vector and non-fixed vectors are a two different things, but again, I think @lemaitre  makes a lot of good points.

**Comment by penzn** [comment]
Thanks @jan-wassenberg. This is a pretty good description of the challenges going forward, closing this issue in favor of #7
